components/brave_wallet/browser/simple_hash_client.cc

```diff
@@ -256,6 +263,264 @@ void SimpleHashClient::OnFetchAllNFTsFromSimpleHash(
+
+  api_request_helper_->Request("GET", url, "", "", std::move(internal_callback),
+                               MakeBraveServicesKeyHeader(),
+                               {.auto_retry_on_network_change = true});
+}
+
+void SimpleHashClient::OnFetchSolCompressedNftProofData(
+    FetchSolCompressedNftProofDataCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode()) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  // Invalid JSON becomes an empty string after sanitization
+  if (api_request_result.value_body().is_none()) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  std::move(callback).Run(
+      ParseSolCompressedNftProofData(api_request_result.TakeBody()));
+}
+
+void SimpleHashClient::GetNftBalances(
+    const std::string& wallet_address,
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    mojom::CoinType coin,
+    GetNftBalancesCallback callback) {
+  if (nft_identifiers.size() > kSimpleHashMaxBatchSize) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  GURL url = SimpleHashClient::GetNftsUrl(coin, nft_identifiers);
+  if (!url.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  auto internal_callback = base::BindOnce(
+      &SimpleHashClient::OnGetNftsForBalances, weak_ptr_factory_.GetWeakPtr(),
+      coin, wallet_address, std::move(nft_identifiers), std::move(callback));
+
+  api_request_helper_->Request("GET", url, "", "", std::move(internal_callback),
+                               MakeBraveServicesKeyHeader(),
+                               {.auto_retry_on_network_change = true});
+}
+
+void SimpleHashClient::OnGetNftsForBalances(
+    mojom::CoinType coin,
+    const std::string& wallet_address,
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    GetNftBalancesCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode() ||
+      api_request_result.value_body().is_none()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  std::optional<
+      base::flat_map<std::string, base::flat_map<std::string, uint64_t>>>
+      owners = ParseBalances(api_request_result.TakeBody());
+
+  if (!owners) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  // For each NFT identifier, create the composite key from the corresponding
+  // chain_id, contract_address, and token_id (if applicable), and look up the
+  // map of owners using the case insensitive composite key. Check if
+  // the wallet_address is in the owners map and add the balance to the
+  // balances vector (keeping the original order).
+  std::vector<uint64_t> balances;
+  for (const auto& nft_identifier : nft_identifiers) {
+    std::string composite_key =
+        nft_identifier->chain_id + "." + nft_identifier->contract_address;
+    if (coin == mojom::CoinType::ETH) {
+      composite_key += "." + nft_identifier->token_id;
+    }
+
```

Comment:
We don't need composite_key if we already have mojom nft_identifier which has `< operator` and allowed to be a key in the map